### ISSUE TITLE
Feature request: final_end_of_line = implicit | explicit

### ISSUE THREAD

**Issue by perennialmind** [editorconfig/editorconfig#394]
Mismatched linebreak conventions (after or between) can be addressed directly. `insert_final_newline` provides for two guaranties for the tail end of files: either some EOL codes or none. It's one way to cope with the mismatch between unix-flavored line **terminators** and windows-flavored line **separators**, but does not address the distinction directly.

Going by the docs, `insert_final_newline = false` might as well be `trim_trailing_newlines = true`. Some (all?) implementations work that way. Combining it with  `trim_trailing_whitespace = true`, you basically get old-school ["virtual space"](https://blogs.msdn.microsoft.com/zainnab/2010/02/28/understanding-virtual-space/). Whitespace as grid positioning, not content.

Delimiter placement (between or after) interacts with the editor's buffer on both open and save, whereas `insert_final_newline` only applies to when saving. Given that, and the confusing construction of the existing option, I propose a new option:

```
final_end_of_line = implicit
final_end_of_line = explicit
```

Editors that normally treat line delimiters as separators a la `notepad`, can maintain the same internal buffer and cope with the foreign convention if they:

1. ignore the final delimiter (if found) **when opening**
2. _and_ append an extra **when saving**.

Editors that presume line terminators a la `vi`, can do the opposite:
1. infer an additional delimiter **when opening**
2. _and_ elide one  **when saving**.

P.S. I'd prefer it to be contextual by eol-value, but don't see that happening.
```
final_end_of_line = implicit-for-crlf explicit-for-lf
```

**Comment by craigbarnes** [comment]
This reads a bit like a brain dump, with a ton of implicit context. It's pretty hard to decipher what you're suggesting here.

**Comment by perennialmind** [comment]
@craigbarnes Sorry I wasn't as clear as I'd thought I was. :smile: I'll try to elaborate on the rationale.

Most editorconfig options relate to the finer points of on-disk representation of a common idiom: the line buffer. Options `tab_width`, `end_of_line`, and `charset` clarify the ambiguities of "plain text". Options  `trim_trailing_whitespace` and `insert_final_newline` specify whether whitespace past the rightmost and bottom-most printable characters are part of content or are apart from it. The Microsoft blog post<sup>`[1]`</sup> I linked to earlier elaborates on that perspective.

There is a separate on-disk representation ambiguity that applies to text editors. In the Unix tradition, every line is terminated with a newline character ("\n" a.k.a. LineFeed). Absent this terminator, the file can be considered malformed, with various surprising consequences. If `foo` is non-empty and doesn't end in a newline, `cat foo bar` or even `cat foo` joins the trailing characters of foo with the first line of bar or the command prompt. Right or wrong, that's the convention. Tools like `git` complain when they encounter non-conformance. Editors like `vim` will automatically "fix" the file by terminating that last line. Which then shows up as a spurious modification to `git`.

Traditional text editors on Windows take the opposite view: the last line is the one **not** followed by "\r\n". Notepad will preserve empty lines at the end of a file, and if the last bytes in the file are "\r\n", then that file _has_ an empty line at the end. Editors of that tradition will never complain about the end of a file, because it can never be "wrong".

Plaintext lines are either **terminated** or **separated**. Put another way, line delimiters go **after** or **between**. I do know of editors with directly applicable settings. Sadly I don't know any that tie those settings to `end_of_line`, which is why I don't see "contextual by eol-value" as realistic.

My main objection is that `insert_final_newline` is relevant to the after/between ambiguity but does not address it directly – and sometimes people expect it to. Witness #269, #393, #337, sindresorhus/editorconfig-sublime#78.  I think addressing it directly is worth doing. :man_shrugging:

`[1]`: https://blogs.msdn.microsoft.com/zainnab/2010/02/28/understanding-virtual-space/

**Comment by craigbarnes** [comment]
> In the Unix tradition, every line is terminated with a newline character ("\n" a.k.a. LineFeed). Absent this terminator, the file can be considered malformed, with various surprising consequences. If `foo` is non-empty and doesn't end in a newline, `cat foo bar` or even `cat foo` joins the trailing characters of foo with the first line of bar or the command prompt.

This seems like more of a Windows-centric rationalization of Unix than something done consciously by Unix software authors. When you ask `cat` to concatenate 2 files, it does exactly as requested -- if a file has no "trailing" newline `cat` doesn't add one, since that wouldn't fit the definition of "concatenation". It's often annoying and undesired for it to behave this way, but I wouldn't call it "unexpected".

Id say the real difference is that Unix has a tradition of treating files/filenames/strings/etc. as "just a bunch of bytes" whereas Windows often implies additional context and special cases to things.

I think the real point should be to make EditorConfig sufficiently configurable to do what people want or expect. The reason people want a trailing newline on Unix isn't tradition but practicality. If there's demand for other options (either because it's what people are used to or because they find it useful) they should be accommodated too.

**Comment by perennialmind** [comment]
Heh. That's what I get for trying to cater to a Windows-centric audience. :smile: I'm one of those folks with a heavily customized prompt: I like that I'm able to `$(cat)` a file into my prompt. I don't actually do that, but I appreciate that I can. Mostly, it's surprising when you use unix-land utilities on a file that came from non-unix-land or tools born there.

Just to be clear, the takeaway I was going for was not that `cat` operates at the byte level. It's that unix lines are _defined_ by a terminating byte sequence. Without it, it's not a line, and the file doesn't qualify as "text file" by POSIX. Really.<sup>`[1][2]`</sup> The record-terminator model is baked into many, many tools. How many local user accounts do I have? Let me just `wc -l /etcpasswd` and I'll tell you. That POSIX line-count utility counts `LF` bytes, with no concern for any characters after the last one. The fact that so many modern editors treat a "trailing newline" as superfluous or signifying an empty line – that represents a conflict with software that doesn't.

People look to `insert_final_newline` to resolve that conflict by specifying and preserving the author's intent, but the option is defined to do something very different. Especially `insert_final_newline = false`, which most editors would treat as a separate option entirely. That's argument №1: fidelity to intent.

I focused on platform tradition because the editors tend to fall on one side or the other. Atom, Visual Studio Code, and SQL Server Management Studio all share a common interpretation of line delimiter semantics, applied to both `LF` and `CRLF` files. There's friction when working with those files and `git` or any other unix-y tools, which impose contrary semantics.

In Vim, I can set `nofixendofline` and `noendofline` when working with Windows-flavored text files. In Visual Studio Code, I can combine `"editor.renderFinalNewline": false` and `"files.insertFinalNewline": true` when working with POSIX-flavored files. To my earlier point, Visual Studio Code `"insertFinalNewline": false` is very, _very_ different from editorconfig insert_final_newline = false. That editorconfig setting maps to `"files.trimFinalNewlines": true`. I'd like an editorconfig option to trigger the first group of settings, and I don't think I'm alone. Argument №2: practical and wanted – I think – by users.

P.S. I think Microsoft introduced the text/binary hack to accommodate the unix-y assumptions built into programming environments like C. Mostly, they paid a heavy price for actually adopting to the IETF-standardized<sup>`[3]`</sup> NVT line encoding<sup>`[4]`</sup>, the earliest platform-neutral formal mapping of ASCII values to the "new line" concept of which I'm aware. But I digress. :smile:

`[1]`. http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206
`[2]`. https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xbd_chap03.html#tag_21_03_00_83
`[3]`. https://www.rfc-editor.org/old/EOLstory.txt
`[4]`. https://www.ietf.org/rfc/rfc854.txt

**Comment by craigbarnes** [comment]
Yeah, I can definitely see your point. Your first comment makes a lot more sense to me now.

It seems like the `final_end_of_line` option would solve the problem for those who understand both perspectives, but I imagine it would probably confuse people who gravitate to one or the other.

Do you think the option could be distilled into something more implicit (or more aptly named at least) that people would grok on sight, instead of requiring a knowledge of the subtleties? I find many people even struggle to understand the `eol` and `text` options in `.gitattributes` files.

**Comment by perennialmind** [comment]
I get where you're coming from. `final_end_of_line` made sense to me at the time: I wanted to focus on the semantics rather than any particular expression. That seemed to be where `insert_final_newline` went awry.
 
I'm just going to throw out a few alternatives, one per comment with the same template.

**Comment by perennialmind** [comment]
Option №2

# Windows linebreak
`crlf_starts_new_line = true`

# RFC-style line endings
`crlf_starts_new_line = false`

# POSIX (Linux) line endings
`lf_starts_new_line = false`

# Editor default
`crlf_starts_new_line = unset`
`lf_starts_new_line = unset`

**Comment by perennialmind** [comment]
Option №3

Windows linebreak
`crlf_placement = between-lines`

RFC-style line endings
`crlf_placement = after-lines`

POSIX (Linux) line endings
`lf_placement = after-lines`

**Comment by perennialmind** [comment]
Option №4

Windows linebreak
`crlf_after = all-but-final-line`

RFC-style line endings
`crlf_after = every-line`

POSIX (Linux) line endings
`lf_after = every-line`

**Comment by perennialmind** [comment]
Option №5

Windows linebreak
`last_crlf_line = after-last-crlf`

RFC-style line endings
`last_crlf_line = before-last-crlf`

POSIX (Linux) line endings
`last_lf_line = before-last-lf`

**Comment by perennialmind** [comment]
Option №6

Windows linebreak
`final_crlf_line = linetext-nocrlf-eof`

RFC-style line endings
`final_crlf_line = linetext-crlf-eof`

POSIX (Linux) line endings
`final_lf_line = linetext-lf-eof`

**Comment by perennialmind** [comment]
Option №7 (last one, I promise)

Windows linebreak
crlf_effect = ends-line-and-begins-next

RFC-style line endings
crlf_effect = completes-line

POSIX (Linux) line endings
lf_effect = completes-line

**Comment by awwright** [comment]
I think there's a simpler way to describe this behavior, that I would like to use instead:

* `newline_style = separated` —Files start with one line, and a line break adds another line.
* `newline_style = terminated` —Files can have zero or more lines, and each line has any number of non-newline characters, terminated by a newline (`␍␊` or `␊`).

This option would affect how editors display files: a value of `terminated` would not show the extra blank line at an end of a file (since the final ␊ isn't having the effect of starting a new line, it's _only_ terminating the current line). Additionally, `newline style = terminated`, then `insert_final_newline` would be forced to `true`.

Internet and Unix text files (and program output) is specified to be newline-terminated. Many to most editors assume documents are newline-delimited.

**Comment by perennialmind** [comment]
@awwright I agree that the **separate** and **terminate** terms distill the distinction nicely. Personally, I would find your concise option values `separated` and `terminated` intuitive. I proposed a bunch of alternatives because I could no longer tell what would be considered "intuitive". :smile:

If this language were adopted, I'd suggest either the stative verb forms (`separates` and `terminates`) or derived agent nouns (`separator` and `terminator`). As written, the past participle seems to confuse subject and indirect object.

With regard to the option name, I'd prefer something closer to those of Option №7 (`crlf_effect` and `lf_effect`). I see a strong argument for coupling the semantics with the delimiter values. The vast majority of relevant text files are delimited by either windows line-breaks or posix line-endings. The smattering of RFC-style line-endings and deliberately unterminated unix character files constitute corner cases. Using two separate options allows for a way to request the sane behavior for both simultaneously.

Case in point: I have a git repository mixing text files that require `lf` delimiters<sup>`[1]`</sup> with those that require `crlf` delimiters<sup>`[2]`</sup> and those that can be converted to comply with the native platform format. The "correct" settings for those files would mandate POSIX behavior for the first and establish an arbitrary but consistent behavior for the second. Given git's simplistic end-of-line conversion, the third case requires the `terminate` option value for both `CRLF` and `LF` files.

Besides, if you put the delimiter values in the option name, you don't have to choose which one of the two already inconsistent terms in use (by `end_of_line` and `insert_final_newline`) for your new option to be inconsistent. :smile:

I agree that `terminated` or `completes-line` or whatever should interact with `insert_final_newline`, but I think it would be cleaner to say that it _overides_ that option. I wouldn't want to break the clean-up removal of blank lines. The feature is fine, it's just tied to an unfortunate name and needs clarification. How about just redefining it to only apply to contiguous `end_of_line` delimiters _after_ the last non-empty line is fully expressed? Either that or introduce yet another option for that behavior, for completeness.


`[1]` shell scripts with embedded here-docs
`[2]` powershell scripts with authenticode signatures that don't survive line or character-encoding conversion

**Comment by maple3142** [comment]
If I am understanding this correctly, this is only meant to control how editor display the final newline and would not affect actual file content?
There will be 4 cases:

|                | one line file without EOL | one line file end with EOL|
|-----------|-----|------|
| separator | display a single line | display two lines and the second line is empty  |
| terminator | display a single line (with warning saying the file have incomplete line?) | display a single line |
